### ISSUE TITLE
[BUG] new instance with invalid parameters problem

### ISSUE THREAD

**Issue by SkibidiProduction** [NumPower/numpower#58]
When running the following code:

```php
new NDArray();
```

I expect to receive:

```
Fatal error: Uncaught ArgumentCountError: NDArray::__construct() expects exactly 1 argument, 0 given
```

Instead I get a segmentation fault:

```
Segmentation fault (core dumped)
```

The same problem occurs if I try to pass an empty array  or invalid value:

```php
new NDArray([]);
//Segmentation fault (core dumped)

new NDArray('string');
//Segmentation fault (core dumped)
```

If we talk about the problem of passing an empty array, then in #54 and #56 I processed this case and threw an exception. However, now I think that perhaps this is not quite the right approach and we should give the user the opportunity to create instances of the class from an empty array.

**Comment by henrique-borba** [comment]
After testing with valgrind, I identified that the error occurred when calling the destructor. It appears that PHP calls the destructor before dealing with `__construct` method exceptions. The destructor was trying to free an NDArray that did not exist.

Fixed by https://github.com/NumPower/numpower/pull/59

I agree with you that we should allow the creation of an empty array, I will resolve this in an MR separate from the destructor fix and leave this issue open for now.
